### DISCUSSION TITLE
Supabase Auth and Next.js SSR: User redirected back to my website after logging in sees the logged out version of the page.

### DISCUSSION THREAD

**Discussion by lumenwrites** [supabase/supabase#7495] [Questions]
Hi! I am using Next.js and Supabase Auth to build a platform for selling courses.

When the user is logged in and has bought the course, I want to show them the content. When the user is logged out, I want to show them the paywall. It is important to me that course content is server-side rendered.
 
I am checking whether the user is logged (and whether they own the course) inside `getServerSideProps()` like so:

```javascript
import { getUser } from '@supabase/auth-helpers-nextjs'
import prisma from 'prisma/prismaClient'

export async function getServerSideProps({ params, req, res }) {
  // Check whether the user owns this course
  let ownsCourse = false
  const { user } = await getUser({ req, res }) // Access the user object
  if (user) {
    const profile = await prisma.profile.findUnique({
      where: { id: user.id },
      include: { courses: true },
    })
    ownsCourse = profile.courses.some((c) => c.courseId === courseSlug)
  }
  [...]
}
```

My problem is that when the user logs in, and is redirected back to my website, they still see the logged out version of the page (with the paywall). They have to reload the page manually to see the signed in version.

I'm guessing that happens because after the user is redirected back to my website, `getServerSideProps()` sees that they don't have any auth cookies yet, so it renders the signed out version of the page. Then the frontend code realizes that I just logged in, and sets the auth cookies, but at this point the page has already been rendered.

Can you please share some advice? How can I solve this?

You can see my code [here](https://github.com/lumenwrites/gdschool/blob/main/pages/course/%5B...slug%5D.tsx).

**Comment by mansueli** [comment]
Maybe you can [force a refresh](https://flaviocopes.com/nextjs-force-page-refresh/) right when the authenticated user is landing there?

**Reply by lumenwrites** [reply]
Yeah, forcing refresh would work, the problem is figuring out how to do this the moment the user is redirected to the app the first time after they sign in. I followed some advice in [this thread](https://github.com/supabase-community/auth-helpers/issues/150), and managed to make this code to work:

```javascript
    supabaseClient.auth.onAuthStateChange(async (event, session) => {
      if (event === 'SIGNED_IN' && session) {
        setTimeout(() => {
          window.location.href = location.protocol + '//' + location.host + location.pathname
        }, 400)
      }
    })
```

When the user returns from being authenticated, I force refresh the page. Unfortunately it doesn't work without `setTimeout()`, and I don't understand why. Like for some reason it can't manage to set the cookies fast enough before the page reloads. And I know this can't be the right way to do things, I can't use this code in production.

Can you please help me find the correct solution to this problem, or ask someone who has dealt with this kind of thing before?

I can't be the only person struggling with this, any Supabase user who is server-side rendering their pages will encounter this problem, there must be some solution...

**Reply by mansueli** [reply]
You can use the [referer](https://stackoverflow.com/a/63235122/2188186) to see where the user is coming from and use that info to determine whether a refresh is needed. 

I'll also ask someone more familiar with NextJS in the team to check this thread.

**Reply by lumenwrites** [reply]
Referer can be used inside `getServerSideProps()`, but the whole problem is that I have to first load the page in order for Supabase client code to work, and do whatever it does to set the cookies.

**Reply by hf** [reply]
Unfortunately the way the libraries work now the cookies won't be set with the request bringing them back to your site. Some JavaScript has to run to talk to your project's authentication server so that the correct JWTs will be issued. Therefore on the server side you won't be able to know what the user is, until the authentication logic completes. You will be notified of this in the `onAuthStateChanged` event.

Forcing a refresh is the recommended way of doing this if you must use server-side rendering in your application's logic. It is likely not necessary to have a 400 ms timeout there, as the refresh should happen on the next event loop tick, which you can do by setting the timeout to be 0 ms.

You may also be able to handle the login of the user in another page which, when the correct cookies and JWT are issued, you will redirect to the server-side rendered page.

**Reply by lumenwrites** [reply]
Unfortunately, it doesn't work without the 400 ms timeout, it doesn't even work with 100 ms timeout, I really don't understand why.

> You may also be able to handle the login of the user in another page which, when the correct cookies and JWT are issued, you will redirect to the server-side rendered page.

Unfortunately I can't do that, because after the user is logged in, I need them to return back to the page they came from. I can't redirect them to some other page, because then it won't know where to redirect the user back.
